observability/src/main/java/io/grpc/observability/interceptors/GcpLogHelper.java

```diff
@@ -0,0 +1,430 @@
+  private static final Logger logger = Logger.getLogger(GcpLogHelper.class.getName());
+  static final Metadata.Key<byte[]> STATUS_DETAILS_KEY =
+      Metadata.Key.of(
+          "grpc-status-details-bin",
+          Metadata.BINARY_BYTE_MARSHALLER);
+
+  /**
+   *  Class for proto related helper methods.
+   */
+  static final class GcpLogSinkWriter {
+    private final Sink sink;
+    private final TimeProvider timeProvider;
+
+    GcpLogSinkWriter(Sink sink, TimeProvider timeProvider) {
+      this.sink = sink;
+      this.timeProvider = timeProvider;
+    }
+
+    GrpcLogRecord.Builder createTimestamp() {
+      long nanos = timeProvider.currentTimeNanos();
+      return GrpcLogRecord.newBuilder().setTimestamp(Timestamps.fromNanos(nanos));
+    }
+
+
+    /**
+     * Logs the client header.
+     */
+    void logClientHeader(
+        long seqId,
+        String serviceName,
+        String methodName,
+        @Nullable String authority,
+        @Nullable Duration timeout,
+        Metadata metadata,
+        GrpcLogRecord.EventLogger eventLogger,
+        String rpcId,
+        @Nullable SocketAddress peerAddress) {
+      checkNotNull(serviceName, "serviceName");
+      checkNotNull(methodName, "methodName");
+      checkNotNull(rpcId, "rpcId");
+      checkArgument(
+          peerAddress == null || eventLogger == GrpcLogRecord.EventLogger.LOGGER_SERVER,
+          "peerAddress can only be specified by server");
+
+      PayloadBuilder<GrpcLogRecord.Metadata.Builder> pair = createMetadataProto(metadata);
+      GrpcLogRecord.Builder logEntryBuilder = createTimestamp()
+          .setSequenceId(seqId)
+          .setServiceName(serviceName)
+          .setMethodName(methodName)
+          .setEventType(EventType.GRPC_CALL_REQUEST_HEADER)
+          .setEventLogger(eventLogger)
+          .setLogLevel(LogLevel.LOG_LEVEL_DEBUG)
+          .setMetadata(pair.proto)
+          .setPayloadSize(pair.size)
+          .setRpcId(rpcId);
+      if (authority != null) {
+        logEntryBuilder.setAuthority(authority);
+      }
+      if (timeout != null) {
+        logEntryBuilder.setTimeout(timeout);
+      }
+      if (peerAddress != null) {
+        logEntryBuilder.setPeerAddress(socketToProto(peerAddress));
+      }
+      logger.log(Level.FINE, "Writing Request Header to GcpLogSink");
+      sink.write(logEntryBuilder.build());
+    }
+
+    /**
+     * Logs the server header.
+     */
+    void logServerHeader(
+        long seqId,
+        String serviceName,
+        String methodName,
+        Metadata metadata,
+        GrpcLogRecord.EventLogger eventLogger,
+        String rpcId,
+        @Nullable SocketAddress peerAddress) {
+      checkNotNull(serviceName, "serviceName");
+      checkNotNull(methodName, "methodName");
+      checkNotNull(rpcId, "rpcId");
+      checkArgument(
+          peerAddress == null || eventLogger == GrpcLogRecord.EventLogger.LOGGER_CLIENT,
+          "peerAddress can only be specified for client");
```

Comment:
I don't really know if this is true. serverheader is responseheader (easier to think that way) and AFAIK the peerAddress is available on both ends (server and client). server because it is going to go back on the same connection the request came on and client because it received the header from the peer. Unless you know the logic for this check